src/snowflake/connector/aio/_s3_storage_client.py

```diff
@@ -0,0 +1,405 @@
+#
+# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
+#
+
+from __future__ import annotations
+
+import re
+import xml.etree.ElementTree as ET
+from datetime import datetime, timezone
+from io import IOBase
+from logging import getLogger
+from typing import TYPE_CHECKING, Any, NamedTuple
+
+import aiohttp
+
+from ..compat import quote, urlparse
+from ..constants import (
+    HTTP_HEADER_CONTENT_TYPE,
+    HTTP_HEADER_VALUE_OCTET_STREAM,
+    FileHeader,
+    ResultStatus,
+)
+from ..encryption_util import EncryptionMetadata
+from ..s3_storage_client import SnowflakeS3RestClient as SnowflakeS3RestClientSync
+from ._storage_client import SnowflakeStorageClient as SnowflakeStorageClientAsync
+
+if TYPE_CHECKING:  # pragma: no cover
+    from ._file_transfer_agent import SnowflakeFileMeta, StorageCredential
+
+logger = getLogger(__name__)
+
+META_PREFIX = "x-amz-meta-"
+SFC_DIGEST = "sfc-digest"
+
+AMZ_MATDESC = "x-amz-matdesc"
+AMZ_KEY = "x-amz-key"
+AMZ_IV = "x-amz-iv"
+
+ERRORNO_WSAECONNABORTED = 10053  # network connection was aborted
+
+EXPIRED_TOKEN = "ExpiredToken"
+ADDRESSING_STYLE = "virtual"  # explicit force to use virtual addressing style
+UNSIGNED_PAYLOAD = "UNSIGNED-PAYLOAD"
+
+RE_MULTIPLE_SPACES = re.compile(r" +")
+
+
+class S3Location(NamedTuple):
+    bucket_name: str
+    path: str
+
+
+class SnowflakeS3RestClient(SnowflakeStorageClientAsync, SnowflakeS3RestClientSync):
+    def __init__(
+        self,
+        meta: SnowflakeFileMeta,
+        credentials: StorageCredential,
+        stage_info: dict[str, Any],
+        chunk_size: int,
+        use_accelerate_endpoint: bool | None = None,
+        use_s3_regional_url: bool = False,
+    ) -> None:
+        """Rest client for S3 storage.
+
+        Args:
+            stage_info:
+        """
+        SnowflakeStorageClientAsync.__init__(
+            self,
+            meta=meta,
+            stage_info=stage_info,
+            chunk_size=chunk_size,
+            credentials=credentials,
+        )
+        # Signature version V4
+        # Addressing style Virtual Host
+        self.region_name: str = stage_info["region"]
+        # Multipart upload only
+        self.upload_id: str | None = None
+        self.etags: list[str] | None = None
+        self.s3location: S3Location = (
+            SnowflakeS3RestClient._extract_bucket_name_and_path(
+                self.stage_info["location"]
+            )
+        )
+        self.use_s3_regional_url = use_s3_regional_url
+        self.location_type = stage_info.get("locationType")
+
+        # if GS sends us an endpoint, it's likely for FIPS. Use it.
+        self.endpoint: str | None = None
+        if stage_info["endPoint"]:
+            self.endpoint = (
+                f"https://{self.s3location.bucket_name}." + stage_info["endPoint"]
+            )
+        # self.transfer_accelerate_config(use_accelerate_endpoint)
+        self.transfer_accelerate_config(False)
+        # TODO: fix accelerate logic
```

Comment:
have you created a JIRA for this? if so let's also paste the jira number in the comment